File: cpp/lib/extend/extend.cpp
#include "extend.hpp"
#include "smith_waterman.hpp"
#include "../../util/display.hpp"

namespace Blastn {

#define MIN(v1, v2) (((v1) > (v2)) ? (v2) : (v1))
#define MAX(v1, v2) (((v1) < (v2)) ? (v2) : (v1))

/* 
 *        ex) Query:     G'TC'TGAA'CT'GAGC
 *            Subject:  AG'TC'TGATGA'CT'GGGGAACTCGA
 *            Left Word:  'TC'
 *            Right Word: 'CT'
 * 
 *            Steps:
 *          1. Query:    G'TC'                 -> score high enough? (extend left)
 *             Subject:  G'TC'
 *          2. Query:    G'TC'T                -> score high enough? (extend right)
 *             Subject:  G'TC'T
 *          3. Query:    G'TC'TGAA             -> score high enough? (extend right more...)
 *             Subject:  G'TC'TGAT
 *          4. Query:    G'TC'TGAA--           -> score high enough? (add gaps)
 *             Subject:  G'TC'TGATGA
 *          5. Query:    G'TC'TGAA--'CT'       -> score high enough? (add right word)
 *             Subject:  G'TC'TGATGA'CT'
 *          6. Query:    G'TC'TGAA--'CT'GAGC   -> score high enough? (extend right more...)
 *             Subject:  G'TC'TGATGA'CT'GGGG
 *          7. Done!
 */

static inline s32 sw_score(string qextended, string sextended, s32 match, s32 mismatch, s32 gap, u32 flag)
{
    switch (flag) {
        case SW::FPGA:
            if (qextended.size() <= SW_MAX_BYTES * 4)
                return smith_waterman_fgpa(qextended.c_str(), sextended.c_str(), qextended.size());
        case SW::PRESERVE_MEM:
            return smith_waterman_preserve(qextended.c_str(), sextended.c_str(), match, mismatch, gap, sextended.size(), qextended.size());
        case SW::NO_PRESERVE_MEM:
            return smith_waterman_no_preserve(qextended, sextended, match, mismatch, gap);
        case SW::MULTI_THREAD:
            return smith_waterman_mt(qextended, sextended, match, mismatch, gap);
        default:
            std::cerr << "Error: INVALID flag for Extending " << flag << std::endl;
            std::exit(-1);
    }
}

Extended extend_and_score(AdjacentPair pair,
                          string& query,
                          string& subject,
                          s32 match,
                          s32 mismatch,
                          s32 gap,
                          f32 ratio,
                          bool score,
                          bool printing,
                          u32 flag)
{
    // find left most indices
    u32 sleftindex  = MIN(pair.sindex1, pair.sindex2);
    u32 qleftindex  = MIN(pair.qindex1, pair.qindex2);
    u32 srightindex = MAX(pair.sindex1, pair.sindex2);
    u32 qrightindex = MAX(pair.qindex1, pair.qindex2);

    // build string
    string qextended = query.substr(qleftindex, pair.length);
    string sextended = subject.substr(sleftindex, pair.length);

    // running Smith Waterman score
    s32 running_score = 0;

    // extend left
    s32 qexindex = qleftindex;
    s32 sexindex = sleftindex;
    while (qexindex - 1 >= 0 && sexindex - 1 >= 0) {
        qexindex--; sexindex--;
        qextended = query[qexindex] + qextended;
        sextended = subject[sexindex] + sextended;

        if (score) {
            running_score = sw_score(qextended, sextended, match, mismatch, gap, flag);

            if (running_score < ratio * qextended.size() * match)
                return Extended{ INVALID, 0, 0, 0 };
        }
    }

    // the left-most index in the subject
    u32 sindex = (u32)sexindex;
    u32 qindex = (u32)qexindex;
    
    // extend left pair to the right
    qexindex = qleftindex + pair.length - 1;
    sexindex = sleftindex + pair.length - 1;
    while ((u32)qexindex + 1 < qrightindex && (u32)sexindex + 1 < srightindex) {
        qexindex++; sexindex++;
        qextended = qextended + query[qexindex];
        sextended = sextended + subject[sexindex];
    }

    // extend right with gaps until qextended aligns with subject
    while ((u32)sexindex + 1 < srightindex) {
        qexindex++; sexindex++;
        qextended = qextended + STR_GAP;
        sextended = sextended + subject[sexindex];
    }

    // append the right pair
    qextended = qextended + query.substr(qrightindex, pair.length);
    sextended = sextended + subject.substr(srightindex, pair.length);

    // extend right
    qexindex = qrightindex + pair.length - 1;
    sexindex = srightindex + pair.length - 1;
    while ((s32)(qexindex + 1) < (s32)query.size() && (s32)(sexindex + 1) < (s32)subject.size()) {
        qexindex++; sexindex++;
        qextended = qextended + query[qexindex];
        sextended = sextended + subject[sexindex];

        if (score) {
            running_score = sw_score(qextended, sextended, match, mismatch, gap, flag);

            if (running_score < ratio * qextended.size() * match)
                return Extended{ INVALID, 0, 0, 0 };
        }
    }

    if (printing) {
        std::cout << "Subject Ext:\t" << sextended << std::endl;
        std::cout << "Quer Ext:\t" << qextended << std::endl;
    }
    return Extended{ qextended, sindex, qindex, running_score };
}

ExtendedSequenceMap extend_filter(PairedSequenceMap& pairs,
                                  SequenceMap& query,
                                  SequenceMap& subject,
                                  s32 match,
                                  s32 mismatch,
                                  s32 gap,
                                  f32 ratio,
                                  int flag)
{
    ExtendedSequenceMap result;
    bool found = false;
    Progress progress{ pairs.size() };

    for (auto& sname_quermap : pairs) {
        ExtendedPairsMap temp;
        for (auto& qname_pairvec : sname_quermap.second) {
            for (auto adjacent_pair : qname_pairvec.second) {

                Extended ext = extend_and_score(adjacent_pair,
                                                query[qname_pairvec.first],
                                                subject[sname_quermap.first], // sw?  print? flag?
                                                match, mismatch, gap, ratio,   true, false, flag);
                // the word scored below the minscore
                if (ext.extended_pair == INVALID)
                    continue;
                
                // check to see if the extended pair was inserted yet
                found = false;
                for (auto& e : temp[qname_pairvec.first]) {
                    if (ext.sindex == e.sindex) {
                        found = true;
                        break;
                    }
                }
                // don't record the pair if it was already found
                if (found)
                    continue;

                // vector<Extended> already initialized
                temp[qname_pairvec.first].push_back(ext);
            }
        }
        if (!temp.empty())
            result[sname_quermap.first] = temp;

        progress.update();
    }
    return result;
}

} // Blastn

File: python/lib/extend.py
from collections import defaultdict
from copy import copy
from typing import Dict, List
import tqdm

from .smith_waterman import smith_waterman
from .pairs import AdjacentPair

"""
Internal
"""

class Extended:
    def __init__(self, extended_pair, dindex, qindex):
        """
        @brief: Hold the extended pair and where it was found in the database.
        """
        self.extended_pair = extended_pair
        self.dindex = dindex
        self.qindex = qindex
    
    def __str__(self):
        return str(self.__dict__)
    def __repr__(self):
        return self.__str__()

"""
2 adjacent pairs
LEFT (lesser index)
    extend leftward one at a time and score
MIDDLE
    fill with gap characters
RIGHT (greater index)
    extend rightward one at a time and score
"""

def extend_and_score(pair: AdjacentPair,
                     query: str,
                     data: str,
                     match: int,
                     mismatch: int,
                     gap: int,
                     minscore: int,
                     score=False,
                     printing=True) -> Extended:
    """
    @brief: Extend an adjacent pair while scoring each extension. Stop extending if the extended word \\
            scores too low. \\
    @param pair:     The adjacent pair of words and their indices. \\
    @param query:    The section the adjacent pair is from. \\
    @param data:     The section the adjacent pair is going to be extended to. \\
    @param match:    Smith Waterman score for a match. \\
    @param mismatch: Smith Waterman score for a mismatch. \\
    @param gap:      Smith Waterman score for a gap. \\
    @param minscore: Minimum Smith Waterman score before the extending stops. \\
    @param score:    Perform Smith Waterman scoring on true. \\
    @param printing: Print the data after extending on true. \\
    @return: The string containing both pairs extended to the query from the database and the database index. \\
    """
    # find left-most indices
    dleftindex = min([pair.dindex1, pair.dindex2])
    qleftindex = min([pair.qindex1, pair.qindex2])
    drightindex = max([pair.dindex1, pair.dindex2])
    qrightindex = max([pair.qindex1, pair.qindex2])

    # build string
    qextended = query[qleftindex:qleftindex + pair.length]
    dextended = data[dleftindex:dleftindex + pair.length]

    # extend left
    qexindex = copy(qleftindex)
    dexindex = copy(dleftindex)
    while qexindex - 1 >= 0 and dexindex - 1 >= 0:
        qexindex -= 1
        dexindex -= 1
        qextended = query[qexindex] + qextended
        dextended = data[dexindex] + dextended
        if score:
            s = smith_waterman(qextended, dextended, match=match, mismatch=mismatch, gap=gap, just_score=True)
            if s < minscore:
                return None

    # the left-most index in the database
    dindex: int = copy(dexindex)
    qindex: int = copy(qexindex)
    
    # extend left pair to the right
    qexindex = qleftindex + pair.length - 1
    dexindex = dleftindex + pair.length - 1
    while qexindex + 1 < qrightindex and dexindex + 1 < drightindex:
        qexindex += 1
        dexindex += 1
        qextended = qextended + query[qexindex]
        dextended = dextended + data[dexindex] # OOB issue
    
    # extend right with gaps until qextended aligns with data
    while dexindex + 1 < drightindex:
        qexindex += 1
        dexindex += 1
        qextended = qextended + '-'
        dextended = dextended + data[dexindex]
    
    # append the right pair
    qextended = qextended + query[qrightindex:qrightindex + pair.length]
    dextended = dextended + data[drightindex:drightindex + pair.length]
    
    # extend right
    qexindex = qrightindex + pair.length - 1
    dexindex = drightindex + pair.length - 1
    while qexindex + 1 < len(query) and dexindex + 1 < len(data):
        qexindex += 1
        dexindex += 1
        qextended = qextended + query[qexindex]
        dextended = dextended + data[dexindex]
        if score:
            s = smith_waterman(qextended, dextended, match=match, mismatch=mismatch, gap=gap, just_score=True)
            if s < minscore:
                return None
    
    if printing:
        print(f"Data Ext:\t{dextended}")
        print(f"Quer Ext:\t{qextended}")
    
    return Extended(qextended, dindex, qindex)

"""
External
"""

def extend_filter(pairs: Dict[str, Dict[str, List[AdjacentPair]]],
                  query: Dict[str, str],
                  data: Dict[str, str],
                  minscore: int,
                  match: int,
                  mismatch: int,
                  gap: int) -> Dict[str, Dict[str, List[Extended]]]:
    """
    @brief: Given adjacent pairs, the database and query, extend the pairs from the query to the database.
    @param pairs:    The mapped data of adjacent pairs to extend and filter while extending.
    @param query:    The map of query names to their entire query sequence.
    @param data:     The map of data names to their entire data sequence.
    @param minscore: The minimum smith waterman score allowed before needing to remove the word.
    @param match:    The smith waterman score when two characters are the same.
    @param mismatch: The smith waterman score when two characters are not the same.
    @param gap:      The smith waterman score when there is a gap character.
    @return A map of data names to query names to a list of extended matches with their data base index.
    """
    result: Dict[str, Dict[str, List[Extended]]] = {}
    
    for dname, queries in tqdm.tqdm(pairs.items()):
        temp = defaultdict(list)
        for qname, adjacent_pairs in queries.items():
            for adjacent_pair in adjacent_pairs:
                extended_pair = extend_and_score(
                    adjacent_pair,
                    query[qname],
                    data[dname],
                    match,
                    mismatch,
                    gap,
                    minscore,
                    score=False,
                    printing=False
                )
                # the word scored above minscore and is not None
                if extended_pair:
                    # don't add repeat extended_pairs
                    if not any(extended_pair.dindex == ep.dindex for ep in temp[qname]):
                        temp[qname].append(extended_pair)
        # temp has items in it, so record it
        if temp:
            result[dname] = dict(temp)
    
    return result
